src/main/java/org/swisspush/reststorage/FilePutter.java

```diff
@@ -0,0 +1,162 @@
+
+    /**
+     * Package-private because currently only used internally.
+     */
+    FilePutter(FileSystem fileSystem, String fullPath, FilePutterCallbacks callbacks, Handler<Resource> onCompleteHandler) {
+        this.fileSystem = fileSystem;
+        this.fullPath = fullPath;
+        this.callbacks = callbacks;
+        this.onCompleteHandler = onCompleteHandler;
+    }
+
+    /**
+     * <p>Triggers the configured task. This method should only get called once per
+     * instance!</p>
+     *
+     * @throws IllegalStateException
+     *      Eg. in case method gets called more than one time.
+     */
+    public synchronized void execute() {
+        if( executed ){
+            throw new IllegalStateException( "This putter already got executed." );
+        }
+        this.executed = true;
+        // Setup required context.
+        this.tmpFilePath = "/.tmp/uploads/" + new File(fullPath).getName() + "-" + UUID.randomUUID().toString() + ".part";
+        this.tmpFilePathAbs = callbacks.canonicalize(tmpFilePath);
+        this.tmpFileParentPath = new File(tmpFilePathAbs).getParent();
+        // Prepare directory for temporary file.
+        fileSystem.mkdirs(tmpFileParentPath, result -> {
+            if (result.succeeded()) {
+                openTmpFile();
+            } else {
+                log.warn("Failed to create directory '" + tmpFileParentPath + "'.");
+                resolveWithErroneousResource();
+            }
+        });
+    }
+
+    private void openTmpFile() {
+        fileSystem.open(tmpFilePathAbs, new OpenOptions(), result -> {
+            if (result.succeeded()) {
+                resolveWithTmpFileResource(result.result());
+            } else {
+                log.warn("Failed to open tmp file '{}'.", tmpFilePathAbs);
+                resolveWithErroneousResource();
+            }
+        });
+    }
+
+    private void resolveWithTmpFileResource(final AsyncFile tmpFile) {
+        final DocumentResource d = new DocumentResource();
+        d.writeStream = tmpFile;
+        d.closeHandler = v -> {
+            tmpFile.close(ev -> {
+                moveTmpFileToFinalDestination(d);
+            });
+        };
+        d.addErrorHandler(err -> {
+            log.error("Put file failed:", err);
+            cleanupFile(tmpFile);
+        });
+        // Resolve with ready-to-use resource.
+        onCompleteHandler.handle(d);
+    }
+
+    private void moveTmpFileToFinalDestination(DocumentResource d) {
+        moveToFinalDestinationAttemptCount += 1;
+        // Move/rename our temporary file to its final destination.
+        fileSystem.move(tmpFilePathAbs, fullPath, moveOptions, moveResult -> {
+            if (moveResult.succeeded()) {
+                log.debug("File stored successfully: {}", fullPath);
+                d.endHandler.handle(null);
+            } else if (moveToFinalDestinationAttemptCount < MOVE_TO_FINAL_DESTINATION_ATTEMPT_LIMIT) {
```

Comment:
Should we probably delay the retry by - lets say - 10 millis or so?
And should we give up after a specific time (e.g. 20 seconds) instead of a fix number of repetitions?
Note that we probably have timeouts on http-Requests (e.g. 30 seconds). So if you need 60 seconds for your 256 retries, then the caller probably will never know about a final success.

Reply:
I actually agree with the retry as well as the specific time instead of repetitions.